esp8266-fastled-webserver/Fields.cpp

```diff
@@ -17,479 +17,526 @@
 */
 #include "common.h"
 
-const String NumberFieldType = "Number";
-const String BooleanFieldType = "Boolean";
-const String SelectFieldType = "Select";
-const String ColorFieldType = "Color";
-const String SectionFieldType = "Section";
-const String StringFieldType = "String";
-const String LabelFieldType = "Label";
-const String UtcOffsetIndexFieldType = "UtcOffsetIndex";
-
-uint8_t power = 1;
-uint8_t brightness = brightnessMap[brightnessIndex];
-
-//String setPower(String value) {
-//  power = value.toInt();
-//  if(power < 0) power = 0;
-//  else if (power > 1) power = 1;
-//  return String(power);
-//}
-
-String getPower() {
-  return String(power);
-}
-
-//String setBrightness(String value) {
-//  brightness = value.toInt();
-//  if(brightness < 0) brightness = 0;
-//  else if (brightness > 255) brightness = 255;
-//  return String(brightness);
-//}
-
-// TODO: wrapper class for EEPROM settings, to help ensure clear understanding
-//       of all locations that can cause settings committed / EEPROM write.
-//       Also simplifies reading prior value before writing new value (extend EEPROM life)
-//       Also simplifies later extending CRC, transactional updates, bypassing "dead"
-//       areas of EEPROM, etc.
-
-String getBrightness() {
-  return String(brightness);
-}
-
-String getPattern() {
-  return String(currentPatternIndex);
-}
-
-String getPatterns() {
-  String json = "";
+// only items outside the anonymous namespace can be exported.
+// this keeps the global namespace cleaner, with no runtime/RAM costs.
+inline namespace {
+  enum struct Field_t : uint8_t {
+    Number,
+    Boolean,
+    Select,
+    Color,
+    Section,
+    String,
+    Label,
+    UtcOffset,
+  };
+  constexpr bool IsValid(const Field_t& t) {
+    return (
+      (t == Field_t::String   ) ||
+      (t == Field_t::Label    ) ||
+      (t == Field_t::Color    ) ||
+      (t == Field_t::Boolean  ) ||
+      (t == Field_t::Number   ) ||
+      (t == Field_t::Section  ) ||
+      (t == Field_t::Select   ) ||
+      (t == Field_t::UtcOffset) );
+  }
```

Comment:
if you would add a `Invalid` Field at the end, you could change the IsValid function to

```suggestion
constexpr bool IsValid(const Field_t& t) {
    return (t < Field_t::Invalid);
}
```

Reply:
First, **_THANK YOU_** for the in-depth review!  It's clear that you have spent some time as a developer, and I really appreciate each of these. 
Even if I decide differently, I greatly appreciate the questions so that the alternatives are considered.  Awesome!

I think I understand the design pattern you are suggesting.

One benefit of your design pattern is that it automatically updates when a new enum type is added.  Some drawbacks of your design pattern are that `Invalid` enum values are allowed to exist, it has an unwritten assumption that the enum values are all sequential, a case for `Invalid` must be added to all switch statements, and there's a subtle implication that there is only one invalid value.

FYI, the compiler will optimize the above to a simple check of `< 8`, so the underlying code remains identical.  If we were targeting C++14 or later, this function would be written as a case statement (to automatically warn for missing values).

Both design patterns have benefits and drawbacks, so without a stronger reason, I would prefer to stick with the current function.

Reply:
Oh, but *do* re-open if you feel I should reconsider, and can help me understand additional aspects to consider!